File: cpp_lib/game_runner.cpp
#include "game_runner.h"
#include <iostream>
#include <chrono>
#include <cstring>
#include <boost/functional/hash.hpp>

// Define the empty chunk as a global constant
const Chunk empty_chunk(8, std::vector<bool>(8, false));

// Converts relative position to absolute position
std::pair<int, int> get_abs_position(const std::pair<int, int>& chunk, int x, int y) {
    return {chunk.first * 8 + x, chunk.second * 8 + y};
}

// Converts absolute position to relative position
std::tuple<std::pair<int, int>, int, int> get_rltv_position(int x, int y) {
    return {{x / 8, y / 8}, x % 8, y % 8};
}

// Find the states of all 8 cells surrounding this one
std::pair<std::vector<bool>, Grid> get_srnd_cells(const std::pair<int, int>& pos, bool is_new, const Grid& master_grid) {
    int x = pos.first;
    int y = pos.second;

    std::vector<bool> output;
    Grid new_chunks;

    std::vector<std::pair<int, int>> positions = {{-1, 1}, {0, 1}, {1, 1}, {-1, 0}, {1, 0}, {-1, -1}, {0, -1}, {1, -1}};
    for (const auto& p : positions) {
        auto [chunk, rel_x, rel_y] = get_rltv_position(x + p.first, y + p.second);
        try {
            output.push_back(master_grid.at(chunk)[rel_x][rel_y]);
        } catch (const std::out_of_range&) {
            output.push_back(false);
            auto [cell_chunk, cell_x, cell_y] = get_rltv_position(x, y);
            if (master_grid.at(cell_chunk)[cell_x][cell_y] && !is_new) {
                new_chunks[chunk] = empty_chunk;
            }
        }
    }

    return {output, new_chunks};
}

// Accepts state of a cell along with the states of its 8 neighbors; returns new state for cell
bool cell_next_day(bool cell_state, const std::vector<bool>& surrounding_cells) {
    int live_count = std::count(surrounding_cells.begin(), surrounding_cells.end(), true);
    if (cell_state) {
        return live_count == 2 || live_count == 3;
    } else {
        return live_count == 3;
    }
}

// Accepts full grid and returns the next generation of the grid
Grid next_gen(Grid input_grid) {
    Grid output_grid;
    auto start_time = std::chrono::high_resolution_clock::now();

    // Remove empty chunks from input_grid
    for (auto it = input_grid.begin(); it != input_grid.end();) {
        if (it->second == empty_chunk) {
            it = input_grid.erase(it);
        } else {
            ++it;
        }
    }

    Grid new_chunks;

    for (const auto& [chunk, cells] : input_grid) {
        output_grid[chunk] = Chunk(8, std::vector<bool>(8, false));
        for (int x = 0; x < 8; ++x) {
            for (int y = 0; y < 8; ++y) {
                auto [surrounding_cells, any_new_chunks] = get_srnd_cells(get_abs_position(chunk, x, y), false, input_grid);
                new_chunks.insert(any_new_chunks.begin(), any_new_chunks.end());
                output_grid[chunk][x][y] = cell_next_day(input_grid[chunk][x][y], surrounding_cells);
            }
        }
    }

    for (const auto& [chunk, cells] : new_chunks) {
        input_grid[chunk] = cells;
    }

    for (const auto& [chunk, cells] : new_chunks) {
        output_grid[chunk] = Chunk(8, std::vector<bool>(8, false));
        for (int x = 0; x < 8; ++x) {
            for (int y = 0; y < 8; ++y) {
                auto surrounding_cells = get_srnd_cells(get_abs_position(chunk, x, y), true, input_grid).first;
                output_grid[chunk][x][y] = cell_next_day(input_grid[chunk][x][y], surrounding_cells);
            }
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end_time - start_time;
    std::cout << "Time taken: " << elapsed.count() << " seconds" << std::endl;

    return output_grid;
}

File: lib/game_runner.py
#This program runs John Conway's Game of Life. It accepts a pattern to run, and calculates the next "day."
#The game is run on a grid of cells. Each cell can be either live (True) or dead (False)

import copy, time

empty_chunk = [[False for _ in range(8)] for _ in range(8)]
def create_empty_chunk():
    return [[False for _ in range(8)] for _ in range(8)]
#This is the master dictionary. It contains all open chunks on the grid. A chunk is an 8x8 square of cells on the grid.
#Each chunk is made up of a list matrix of cells.
#An absolute coordinate is one that just has the position of a cell.
#A relative coordinate is the coordinate of a cell within a specific chunk. These below are relative coordinates.
#The relative coordinate of a cell within a chunk does not change based on what quadrant the chunk is in. The bottom
#left cell in a chunk is always (0, 0).
#To retrieve a cell from today_grid (ngInputGrid), use the following format: today_grid[chunk][X][Y]
#The today_grid (ngInputGrid) stores the current state of every cell. The tomorrow_grid (ngOutputGrid) is filled every cycle as the game decides what the
#next day will look like.


#Converts relative position to absolute position. accepts X and Y of chunk, then X and Y of cell in chunk.
def get_abs_position(gapChunk, gapX, gapY):
    return (int(gapChunk[0]) * 8 + int(gapX), int(gapChunk[1]) * 8 + int(gapY))

#Converts absolute position to relative position. Accepts a list with X and Y of cell.
def get_rltv_position(grpX, grpY):
    return ((int(grpX) // 8, int(grpY) // 8), int(grpX) % 8, int(grpY) % 8)

#Find the states of all 8 cells surrounding this one. Accepts absolute position.
def get_srnd_cells(gscXY, gscIsNew, gscMasterGrid):
    (gscX, gscY) = gscXY
    gscOutput = []
    gscNewChunks = {}
    for gscPosition in ((-1, 1), (0, 1), (1, 1), (-1, 0), (1, 0), (-1, -1), (0, -1), (1, -1)): #iterates over all 8 surrounding cells
        #finds the relative position of one of the cells surrounding the chosen cell
        gscOuterCell = get_rltv_position(gscX + gscPosition[0], gscY + gscPosition[1]) #TODO instead of gsc() receiving absolute position, receive relative and do this calculation using modulus and such.
        try:
            gscOutput.append(gscMasterGrid[gscOuterCell[0]][gscOuterCell[1]][gscOuterCell[2]])
        except KeyError:
            gscOutput.append(False)
            gscCell = get_rltv_position(gscX, gscY)
            # Only create new chunk if current cell is live
            if gscMasterGrid[gscCell[0]][gscCell[1]][gscCell[2]] and not gscIsNew:
                gscNewChunks[gscOuterCell[0]] = create_empty_chunk()

    return (gscOutput, gscNewChunks)


#Accepts state of a cell along with the states of its 8 neighbors; returns new state for cell.
def cell_next_day(cndCellState, cndSrndngCells):
    cndLive = cndSrndngCells.count(True)
    if cndCellState == True:
        if cndLive == 2 or cndLive == 3:
            return True
    else: # is dead
        if cndLive == 3:
            return True
    return False




#accepts full grid and returns the next generation of the grid.
def next_gen(ngInputGrid):
    global empty_chunk
    ngOutputGrid = {}
    start_time = time.perf_counter() #test speed #TODO remove for final product

    #remove empty chunks from ngInputGrid
    for chunk in copy.deepcopy(ngInputGrid):
        if ngInputGrid[chunk] == empty_chunk:
            del ngInputGrid[chunk]
            #print('closed chunk: ' + str(chunk))

    ngNewChunks = {}

    # First pass - process existing chunks and identify needed new chunks
    for chunk in ngInputGrid:
        ngOutputGrid[chunk] = []
        for Xcoord in range(8):
            ngOutputGrid[chunk].append([])
            for Ycoord in range(8):
                ngSrndCells, ngAnyNewChunks = get_srnd_cells(get_abs_position(chunk, Xcoord, Ycoord), False, ngInputGrid)
                ngNewChunks |= ngAnyNewChunks
                ngOutputGrid[chunk][Xcoord].append(cell_next_day(ngInputGrid[chunk][Xcoord][Ycoord], ngSrndCells))

    #add newly any newly opened chunks to ngInputGrid
    for addition in ngNewChunks:
        ngInputGrid[addition] = create_empty_chunk()
        #print('opened chunk: ' + str(addition))

    #Iterate over all cells in newly opened chunks:
    for chunk in ngNewChunks: #iterates over every chunk key
        ngOutputGrid[chunk] = []
        for Xcoord in range(8): #iterates over every X coordinate in chunk
            ngOutputGrid[chunk].append([])
            for Ycoord in range(8): #iterates over every Y coordinate
                #find states of surrounding cells
                ngSrndCells = get_srnd_cells(get_abs_position(chunk, Xcoord, Ycoord), True, ngInputGrid)[0] #glitchy with returned index?
                #find new state of cell and add to ngOutputGrid
                ngOutputGrid[chunk][Xcoord].append(cell_next_day(ngInputGrid[chunk][Xcoord][Ycoord], ngSrndCells))

    end_time = time.perf_counter()
    print(end_time - start_time)
    return ngOutputGrid





#Coding tip: Local variables do not have underscores and are prefixed by the name of their domain.

#TODO: Optimization: make loops shorter, e.g. use while instead of for. Remove unecessary variables and asserts (such as the one in get_window(). What else?
#TODO speed test
#TODO give iterators similar names for optimization?
#TODO optimize remove int() from position convert functions
